### PR TITLE
perf(platform): cache the process-table walk on the format render path

### PR DESCRIPTION
**Pull request by Gerrrt**
`#{pane_current_command}` and `#{pane_current_path}` resolve a pane's foreground process, which on Windows means a full
CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS) over every process on the machine (~340 on a normal desktop). Both variables are expanded on the server's per-output render path, so a status bar with `#{pane_current_path}` plus a `set-titles-string` with `#{pane_current_command}` took two full system walks per repaint, on the thread that also delivers keystrokes.

This is a pattern the codebase already agrees with: window_ops's scroll path guards the same calls with a 2s per-pane cache, noting the walk is too expensive to redo per wheel tick. The format path had no equivalent. And the number of independent walkers keeps growing. #491 just added foreground_leaf_name and foreground_is_vt_bridge, so a single Ctrl+C now snapshots the table twice.

Introduce a shared, time-bounded snapshot (`process_table`) and route every walker through it:

- Render-path callers (find_foreground_child_pid) reuse a snapshot up to 250ms old, so a fast-drawing pane pays at most four walks a second instead of one per frame, while a window title still updates fast enough to look instant.
- Input-routing callers: foreground_is_shell / foreground_is_vt_bridge (Ctrl+C) and has_vt_bridge_descendant (mouse transport) pass Duration::ZERO and always enumerate fresh, because acting on a stale process tree would misroute an interrupt or a click. #491's foreground_leaf_name now goes through this too.
- A fresh walk publishes into the shared slot, so a ZERO-age caller warms the cache for a following render-path caller.

Also gate autorename_log behind PSMUX_AUTORENAME_DEBUG. It was the only logger in the tree that was unconditional, doing a file open plus append on this hot path for the first 100 calls of every server; it had left a 700KB file behind.

tests-rs/test_proc_table_cache.rs asserts the property that matters: the number of real walks is bounded by time, not by how many times the render path asks, and that ZERO-age callers always walk. Verified the tests fail when the render TTL is set to zero.

### CODE REVIEW

File: src/platform.rs [hunk 1]
```diff
@@ -2890,8 +2890,23 @@ pub mod process_info {
...
+    /// line. On this developer's machine it had accumulated a 703KB file. The
+    /// 100-entry cap is per-process, so the file grows without bound across
+    /// server restarts and the cap gives no protection against that.
     fn autorename_log(msg: &str) {
         use std::sync::atomic::{AtomicU32, Ordering};
+        use std::sync::OnceLock;
+        static ENABLED: OnceLock<bool> = OnceLock::new();
+        if !*ENABLED.get_or_init(|| {
+            std::env::var("PSMUX_AUTORENAME_DEBUG").unwrap_or_default() == "1"
+        }) {
```
[the remark this answers is not in the document]
Comment (Gerrrt) [reply src/platform.rs:2910]: Fixed in 4921068. The gate now accepts "1" or "true" (case insensitive), matching env_enabled in debug_log.rs so every debug gate in the tree behaves the same way.

File: src/platform.rs [hunk 2]
```diff
@@ -2890,8 +2890,23 @@ pub mod process_info {
...
+            return;
+        }
         static COUNT: AtomicU32 = AtomicU32::new(0);
         let n = COUNT.fetch_add(1, Ordering::Relaxed);
         if n > 100 { return; }
```
[the remark this answers is not in the document]
Comment (Gerrrt) [reply src/platform.rs]: Fixed in 4921068, changed to n >= 100 so it caps at exactly 100 writes. 

One note for context: 
- this cap predates the branch (I only added the env gate around the function). Contradicts the "first 100 entries" doc, so I aligned it while I was in there.

File: src/platform.rs [hunk 3]
```diff
@@ -2977,78 +2992,161 @@ pub mod process_info {
...
+    /// cached.
+    fn process_table(max_age: std::time::Duration) -> Option<ProcTable> {
+        if !max_age.is_zero() {
+            if let Ok(guard) = PROC_TABLE_CACHE.lock() {
+                if let Some((at, table)) = guard.as_ref() {
+                    if at.elapsed() < max_age {
+                        return Some(std::sync::Arc::clone(table));
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Gerrrt) [reply src/platform.rs]: Fixed in 4921068. Both the read and publish sites now recover the inner value with unwrap_or_else(|e| e.into_inner()) instead of dropping the cache when the lock is poisoned. 

The critical sections are only ever a single assignment or a clone, so the recovered value is always a complete entry or None. Never half written.

File: tests-rs/test_proc_table_cache.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+fn walks() -> u64 {
+    PROC_TABLE_WALKS.with(|c| c.get())
+}
+
+/// Force the next render-path call to miss the cache, so a test starts from a
+/// known state regardless of what ran before it.
+fn invalidate() {
+    if let Ok(mut g) = PROC_TABLE_CACHE.lock() {
+        *g = None;
+    }
```
[the remark this answers is not in the document]
Comment (Gerrrt) [reply tests-rs/test_proc_table_cache.rs]: Fixed in 4921068. invalidate() now recovers the poisoned lock the same way the module's lock() helper does, so a panicking test earlier in the run can't leave a populated cache behind and make these order dependent.

**Comment by psmux** [comment]
Merged as 3d288b7. Thanks Garrett, this one is a genuinely good catch and the writeup made it easy to verify.

I did not take the description on faith, so I reproduced everything on clean master first (b3d55f8) before looking at your diff. All three of your claims held up, and one of them was understated.

**The two walks per repaint are real.** Timing each format variable in isolation on master, the cost scales exactly with how many walking variables you ask for:

| format var | p50 |
| --- | --- |
| `#{session_name}` | 12.87ms |
| `#{pane_current_command}` | 26.15ms |
| `#{pane_current_path}` | 25.70ms |
| both together | 38.65ms |

One walk is about 13ms, two are about 26ms. Exactly the shape you described.

**The snapshot is bigger here than you assumed.** Your `~340` was conservative. My `autorename.log` lines read `snapshot_entries=501`, so this box pays even more per walk than yours does.

**The ungated logger had done more damage here than on your machine.** You found a 703KB file. Mine was **2.6MB across 54,802 lines**, accumulated over many server restarts, and a single short benchmark run added another 4,750 bytes with the env var unset. You were right that the per-process 100-entry cap gives no protection across restarts.

After the change, same benchmark, same machine:

| metric | master | merged |
| --- | --- | --- |
| walk overhead, idle | 25.44ms | **0.38ms** |
| walk overhead, under repaint flood | 24.01ms | **0.34ms** |
| autorename.log with debug off | 4,750 bytes | **0** |

The part I spent the most time on was trying to break the correctness of the shared cache, since the slot is process wide while the question it answers is per pane. It holds up. Three panes running different foreground programs report their own command and their own cwd, with no bleed, including 15 interleaved cross pane queries fired inside a single TTL window. Ctrl+C still stops a foreground program (16ms), the vt-bridge mouse path still classifies a `wsl` child correctly, and under a real attached client rendering both variables in the status bar every frame, 827 sustained samples came back at p50 13.57ms, which is the non-walking baseline.

Your `Duration::ZERO` split is the right call and I want to note it specifically, because the lazy version of this patch would have put one global TTL in front of everything and quietly introduced a class of input misrouting bugs that would have been miserable to track down later.

Two things convinced me the 250ms bound is not a parity risk:

1. tmux refreshes these same variables on `status-interval`, which defaults to **15 seconds**. Your bound is 60x fresher than upstream. On Unix tmux also resolves this from the tty foreground pgid rather than enumerating anything, so the whole-table walk is a Windows-only tax we were paying that tmux never pays at all.
2. The precedent you cited is accurate to the line. `window_ops.rs` already ships two 2 second caches in front of this exact call, one commented "too expensive to redo on every wheel tick". Your 250ms is 8x tighter than what was already merged and accepted.

I also checked your claim that the tests fail if the fix is disabled, by setting `RENDER_PATH_TTL` to zero. 4 of the 7 fail. They are real guards, not decoration.

Verification run: full Rust suite **2561 passed, 0 failed**, plus every E2E suite that touches this code (`test_issue299_pane_current_command`, `test_issue495_pane_current_path`, `test_pr469_pane_current_command`, `test_issue266_autorename_override`, `test_issue229_window_name_flash`, `test_issue229_window_name_initial_command`) at 70 assertions and 0 failures. `test_issue277_scroll_controlled` fails 2 assertions, but it fails identically on clean master, so that is pre-existing environment drift on my box and not yours.

One housekeeping note: the branch went stale while it sat, so it conflicted on merge. It was purely two of us adding adjacent rows to the same doc table in `debug_log.rs`, with your `PSMUX_AUTORENAME_DEBUG` row landing next to the `PSMUX_PANE_RAW` row from #502. I kept both and merged it here rather than bounce it back to you for a rebase over a documentation comment.

Good patch. The habit of measuring the thing rather than asserting it, and of making the walk count observable instead of timing it, is exactly what makes this reviewable.
